src/main/java/io/r2dbc/h2/H2ConnectionFactory.java

```diff
@@ -91,4 +156,59 @@ private static ConnectionInfo getConnectionInfo(H2ConnectionConfiguration config
         }
     }
 
+    private static class DefaultCloseableConnectionFactory implements CloseableConnectionFactory {
+
+        private final H2ConnectionConfiguration configuration;
+
+        private final Supplier<SessionClient> clientFactory;
+
+        private volatile SessionClient persistentConnection;
+
+        public DefaultCloseableConnectionFactory(H2ConnectionConfiguration configuration) {
+            this.configuration = configuration;
+            this.clientFactory = () -> getSessionClient(configuration, false);
+            this.persistentConnection = getSessionClient(configuration, true);
+        }
+
+        @Override
+        public Mono<Void> close() {
+            return Mono.defer(() -> {
+
+                SessionClient connection = this.persistentConnection;
+                this.persistentConnection = null;
+
+                if (connection != null) {
+                    return connection.close();
+                }
+
+                return Mono.empty();
+            });
+        }
+
+        @Override
+        public Mono<H2Connection> create() {
+            return Mono.fromSupplier(() -> {
+
+                if (this.persistentConnection == null) {
+                    throw new H2DatabaseExceptionFactory.H2R2dbcNonTransientResourceException(String.format("ConnectionFactory for %s is closed", this.configuration.getUrl()));
+                }
```

Comment:
I'd favor using a suitable import statement to remove `H2DatabaseExceptionFactory`.